alg/teca_gradient.cxx

```diff
@@ -0,0 +1,464 @@
+#include "teca_gradient.h"
+
+#include "teca_cartesian_mesh.h"
+#include "teca_array_collection.h"
+#include "teca_variant_array.h"
+#include "teca_variant_array_impl.h"
+#include "teca_metadata.h"
+#include "teca_array_attributes.h"
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <cmath>
+
+#if defined(TECA_HAS_BOOST)
+#include <boost/program_options.hpp>
+#endif
+
+using std::string;
+using std::vector;
+using std::cerr;
+using std::endl;
+using std::cos;
+
+// #define TECA_DEBUG
+
+namespace {
+
+template <typename num_t>
+constexpr num_t deg_to_rad() { return num_t(M_PI)/num_t(180); }
+
+template <typename num_t>
+constexpr num_t earth_radius() { return num_t(6371.0e3); }
+
+// compute the gradient using second-order, centered finite difference
+// assumes fixed mesh spacing. here we add periodic bc in lon
+// and apply unit stride vector optimization strategy to loops
+template <typename num_t, typename pt_t>
+void gradient(num_t *grad_x, num_t *grad_y, const pt_t *lon, const pt_t *lat,
+    const num_t *f, unsigned long n_lon,
+    unsigned long n_lat, bool periodic_lon=true)
+{
+    size_t n_bytes = n_lat*sizeof(num_t);
+    num_t *delta_x = static_cast<num_t*>(malloc(n_bytes));
```

Comment:
Isn't better to use std::vector<num_t> instead of explicitly malloc?

Reply:
That would be more in the C++ style, yes.  I copied the approach from `teca_vorticity.cxx`, so I'm wondering if there's a performance reason for using a dynamically allocated array rather than a vector: perhaps the overhead from bounds-checking in a vector?  @burlen  - do you have an opinion here on which approach should be used?

Reply:
I ended up going with `std::vector` since the `[]` operator doesn't have any overheads.

Reply:
OK, I'll revert it.

Reply:
@taobrienlbl It's fine  to use vector here, I didn't mean for you to have to revert this, just wanted to give an answer as to why I use malloc.